File: nmos/cert_check.py
# Copyright (C) 2025-2026 Alain Bouchard
# SPDX-License-Identifier: Apache-2.0

"""Startup-time TLS certificate verification.

Helpers invoked during ``nmos_node.py`` startup to fail fast on
mis-configured cert / key / trusted-root-CA combinations.

The two helpers are:

* :func:`check_trusted_ca` — verifies that the leaf cert in a
  trusted-CA bundle chains to the global root CA.
* :func:`check_certificate` — verifies that a server cert chains
  to the global root CA via the intermediates packaged in the
  ``*.chain.pem`` file, that its SAN includes the expected
  per-serial DNS entry, and that the supplied private key
  matches the cert's public key.

Both raise :class:`CertCheckError` on failure (subclass of
``ValueError``); ``nmos_node.py`` translates that into a fatal
``SystemExit`` so the operator gets a single-line, actionable
diagnostic.
"""

from __future__ import annotations

from typing import Any, Sequence

from cryptography import x509
from cryptography.hazmat.primitives.serialization import (
    Encoding,
    PublicFormat,
    load_pem_private_key,
)
from OpenSSL import crypto


class CertCheckError(ValueError):
    """Raised when a startup cert/key/CA validation fails."""


# ---------------------------------------------------------------------------
# PEM loading
# ---------------------------------------------------------------------------

def _load_pem_certs(path: str) -> list[crypto.X509]:
    """Load every ``-----BEGIN CERTIFICATE-----`` block from a PEM file.

    Returns the certs in file order so a ``chain.pem`` file
    (leaf followed by intermediates) yields ``[leaf, *intermediates]``.
    Non-cert PEM blocks (e.g. private keys mixed in) are skipped.
    """
    try:
        with open(path, "rb") as f:
            data = f.read()
    except OSError as exc:
        raise CertCheckError(f"cannot read {path!r}: {exc}") from exc

    certs: list[crypto.X509] = []
    marker = b"-----END CERTIFICATE-----"
    parts = data.split(marker)
    for raw in parts:
        chunk = raw.strip()
        if not chunk or b"-----BEGIN CERTIFICATE-----" not in chunk:
            continue
        full = chunk + b"\n" + marker + b"\n"
        try:
            certs.append(crypto.load_certificate(crypto.FILETYPE_PEM, full))
        except crypto.Error:
            # Mixed file with non-cert blocks — skip and keep walking.
            continue
    if not certs:
        raise CertCheckError(f"no PEM certificates found in {path!r}")
    return certs


# ---------------------------------------------------------------------------
# Chain verification
# ---------------------------------------------------------------------------

def _verify_chain(
    leaf: crypto.X509,
    roots: list[crypto.X509],
    intermediates: list[crypto.X509] | None = None,
) -> None:
    """Run pyOpenSSL's X509StoreContext, mapping ``X509StoreContextError``
    to :class:`CertCheckError` with a clear message."""
    store = crypto.X509Store()
    for r in roots:
        store.add_cert(r)
    ctx = (
        crypto.X509StoreContext(store, leaf, chain=intermediates)
        if intermediates
        else crypto.X509StoreContext(store, leaf)
    )
    try:
        ctx.verify_certificate()
    except crypto.X509StoreContextError as exc:
        raise CertCheckError(f"certificate chain validation failed: {exc}") from exc


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def check_trusted_ca(
    root_ca_paths: str | Sequence[str],
    trusted_ca_path: str,
) -> None:
    """Verify that ``trusted_ca_path``'s leaf cert chains to a cert in
    any of the ``root_ca_paths``.

    Supports multi-root validation: ``root_ca_paths`` may be a single
    path or any iterable of paths. Every PEM cert found across the
    supplied files is loaded into one trust store, and the leaf in
    ``trusted_ca_path`` must chain to one of them.

    Used during nmos_node.py startup whenever ``--nodeTrustedRootCA``
    (or ``--oauth2TrustedRootCA``) is supplied alongside the global
    ``--trustedRootCA``: it asserts that the per-service trusted-CA
    bundle is itself trusted by the global anchor, refusing to start
    otherwise.
    """
    paths: list[str] = (
        [root_ca_paths] if isinstance(root_ca_paths, str) else list(root_ca_paths)
    )
    trusted_certs = _load_pem_certs(trusted_ca_path)
    root_certs: list[crypto.X509] = []
    for p in paths:
        root_certs.extend(_load_pem_certs(p))
    _verify_chain(trusted_certs[0], root_certs)


def check_certificate(
    root_ca_paths: str | Sequence[str],
    cert_path: str,
    key_path: str,
    serial_number: str,
) -> None:
    """Verify a server cert + private key + serial-bound SAN entry.

    Supports multi-root validation: ``root_ca_paths`` may be a single
    path or any iterable of paths. Performs three independent checks:

    1. ``cert_path`` (treated as a ``chain.pem`` — leaf followed by
       intermediates) chains to a cert in any of the ``root_ca_paths``.
    2. The leaf's Subject Alternative Name list includes
       ``f"XYZ-{serial_number}"``. The actual Example-issued certs
       under ``Certificates/build.0/`` use the uppercase ``XYZ-`` prefix
       in their SAN list (verified against ExampleDeviceServer.*.pem),
       so this implementation follows the cert generator's actual
       output.
    3. The supplied private key (``key_path``) matches the leaf cert's
       public key (modulus equality for RSA, public-point equality
       for EC).
    """
    paths: list[str] = (
        [root_ca_paths] if isinstance(root_ca_paths, str) else list(root_ca_paths)
    )
    chain_certs = _load_pem_certs(cert_path)
    root_certs: list[crypto.X509] = []
    for p in paths:
        root_certs.extend(_load_pem_certs(p))

    leaf = chain_certs[0]
    intermediates = chain_certs[1:]
    _verify_chain(leaf, root_certs, intermediates=intermediates)

    # SAN check — uppercase "XYZ-<serial>" matching the certificate
    # generator's actual output.
    expected_dns = f"XYZ-{serial_number}"
    sans = _extract_dns_sans(leaf)
    if expected_dns not in sans:
        raise CertCheckError(
            f"certificate {cert_path!r} SAN does not include "
            f"{expected_dns!r}; found DNS SANs: {sans}"
        )

    # Key↔cert match — load both and compare public-key material.
    try:
        with open(key_path, "rb") as f:
            key_pem = f.read()
    except OSError as exc:
        raise CertCheckError(f"cannot read {key_path!r}: {exc}") from exc

    try:
        key_obj: Any = load_pem_private_key(key_pem, password=None)
    except (ValueError, TypeError) as exc:
        raise CertCheckError(
            f"cannot load private key {key_path!r}: {exc}"
        ) from exc

    # Compare via DER-serialized public-key blobs — works uniformly
    # across RSA / EC / Ed25519 / Ed448 / X25519 / X448 without
    # needing per-algorithm type narrowing.
    leaf_pubkey = leaf.to_cryptography().public_key()
    leaf_der = leaf_pubkey.public_bytes(
        encoding=Encoding.DER,
        format=PublicFormat.SubjectPublicKeyInfo,
    )
    key_der = key_obj.public_key().public_bytes(
        encoding=Encoding.DER,
        format=PublicFormat.SubjectPublicKeyInfo,
    )
    if leaf_der != key_der:
        raise CertCheckError(
            f"private key {key_path!r} does not match certificate {cert_path!r}"
        )


# ---------------------------------------------------------------------------
# Internals
# ---------------------------------------------------------------------------

def _extract_dns_sans(cert: crypto.X509) -> list[str]:
    """Return the DNS-name SAN entries from a cert (empty list if absent)."""
    crypt_cert = cert.to_cryptography()
    try:
        ext = crypt_cert.extensions.get_extension_for_class(
            x509.SubjectAlternativeName,
        )
    except x509.ExtensionNotFound:
        return []
    return list(ext.value.get_values_for_type(x509.DNSName))


def cert_dns_identities(cert_path: str) -> list[str]:
    """Return the DNS names by which a TLS server cert may be identified —
    its Subject CN (when present and parseable as a DNS name) plus every
    DNS-type Subject Alternative Name, in declaration order, with
    duplicates removed.

    Used to populate ``node.tls_server_cert_names`` so the IS-10
    audience check (``nmos.oauth2.aud_entry_allows_current_node``)
    can verify that a token's ``aud`` entry corresponds to one of
    this Node's cert identities — required by IS-10 / "NMOS With
    OAuth2.0" §"Validation" (the ``aud`` DNS name MUST be a CN or
    SAN of the TLS server cert).

    Returns an empty list when the file is missing/unreadable or
    contains no certs — the caller should treat that as "no TLS
    identity available", which causes the audience check to fail
    closed (correct behaviour for unconfigured production).
    """
    try:
        certs = _load_pem_certs(cert_path)
    except CertCheckError:
        return []
    if not certs:
        return []
    leaf = certs[0]
    out: list[str] = []
    seen: set[str] = set()

    # CN — modern certs sometimes omit it (relying on SAN), but when
    # it's present and is a DNS name, it counts as an identity.
    crypt_cert = leaf.to_cryptography()
    for attr in crypt_cert.subject:
        if attr.oid == x509.NameOID.COMMON_NAME:
            cn = str(attr.value)
            if cn and cn not in seen:
                out.append(cn)
                seen.add(cn)

    # DNS-type SANs.
    for name in _extract_dns_sans(leaf):
        if name and name not in seen:
            out.append(name)
            seen.add(name)
    return out

File: rust/crates/nmos-registry-bin/src/cert_check.rs
// Copyright (C) 2025-2026 Alain Bouchard
// SPDX-License-Identifier: Apache-2.0

//! Refusing to start on a mis-configured certificate set.
//!
//! Port of `nmos_registry.py`'s `validate_startup_certs` and the two checks it
//! calls from `nmos/cert_check.py`. It runs after the command line is parsed
//! and before anything binds, "so the operator gets a clear diagnostic instead
//! of a TLS handshake failure at the first connection".
//!
//! # Why this is not merely a nicety
//!
//! Without it, `--registryCertificate` left out means [`crate::listen`]'s
//! `context_for` warns and serves **plain HTTP** -- a deployment that believes
//! it is running TLS and is not. `_server_context`'s warning path exists for
//! that reason and is unreachable in the real program precisely because this
//! check has already exited. Porting one without the other turns a fail-fast
//! into a silent downgrade, which is why this file exists.
//!
//! # The rule about per-interface anchors
//!
//! A per-interface trust anchor must itself chain to the global
//! `--trustedRootCA`. `validate_startup_certs` puts it plainly: "a mis-issued
//! anchor would otherwise be discovered only when a legitimate client was
//! rejected". An anchor nobody verified is an anchor that might trust the wrong
//! issuer.

use std::path::Path;

use openssl::stack::Stack;
use openssl::x509::store::X509StoreBuilder;
use openssl::x509::{X509, X509StoreContext};

use crate::cli::{self, Args};

/// A configuration that cannot be served.
///
/// Every message is prefixed `CONFIG:` and phrased as `validate_startup_certs`
/// phrases it, so an operator reading one implementation's diagnostic can search
/// the other's source for it.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ConfigError(pub String);

impl std::fmt::Display for ConfigError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.0)
    }
}

impl std::error::Error for ConfigError {}

/// Format a string as Python's `!r` does.
///
/// Every one of these messages interpolates a path with `!r`, which yields
/// single quotes where Rust's `{:?}` yields double. The wording is meant to be
/// searchable across both implementations, so an operator who greps one
/// source for a diagnostic they saw from the other finds it -- which stops
/// working the moment the quoting differs.
///
/// Python switches to double quotes only when the value contains a single quote
/// and no double quote; that rule is followed here rather than approximated.
fn repr(value: &str) -> String {
    if value.contains('\'') && !value.contains('"') {
        format!("\"{value}\"")
    } else {
        format!("'{}'", value.replace('\'', "\\'"))
    }
}

/// Format a list of strings as Python's `!r` does: `['a', 'b']`.
fn repr_list(values: &[String]) -> String {
    let inner: Vec<String> = values.iter().map(|value| repr(value)).collect();
    format!("[{}]", inner.join(", "))
}

/// Read every `BEGIN CERTIFICATE` block, in file order.
///
/// Order matters: a `chain.pem` is leaf first, then intermediates.
fn load_certs(path: &Path) -> Result<Vec<X509>, ConfigError> {
    let pem = std::fs::read(path).map_err(|error| {
        ConfigError(format!(
            "CONFIG: cannot read {}: {error}",
            repr(&path.to_string_lossy())
        ))
    })?;
    let certs = X509::stack_from_pem(&pem).map_err(|error| {
        ConfigError(format!(
            "CONFIG: cannot parse {}: {error}",
            repr(&path.to_string_lossy())
        ))
    })?;
    if certs.is_empty() {
        return Err(ConfigError(format!(
            "CONFIG: {} holds no certificate",
            repr(&path.to_string_lossy()),
        )));
    }
    Ok(certs)
}

/// Every certificate across a set of PEM files, as one trust store.
fn load_roots(paths: &[impl AsRef<Path>]) -> Result<Vec<X509>, ConfigError> {
    let mut roots = Vec::new();
    for path in paths {
        roots.extend(load_certs(path.as_ref())?);
    }
    Ok(roots)
}

/// Verify `leaf` chains to one of `roots`, through `intermediates`.
fn verify_chain(leaf: &X509, roots: &[X509], intermediates: &[X509]) -> Result<(), ConfigError> {
    let failed =
        |error: String| ConfigError(format!("certificate chain validation failed: {error}"));

    let mut store = X509StoreBuilder::new().map_err(|e| failed(e.to_string()))?;
    for root in roots {
        store
            .add_cert(root.clone())
            .map_err(|e| failed(e.to_string()))?;
    }
    let store = store.build();

    let mut chain = Stack::new().map_err(|e| failed(e.to_string()))?;
    for intermediate in intermediates {
        chain
            .push(intermediate.clone())
            .map_err(|e| failed(e.to_string()))?;
    }

    let mut context = X509StoreContext::new().map_err(|e| failed(e.to_string()))?;
    let verified = context
        .init(&store, leaf, &chain, |context| context.verify_cert())
        .map_err(|e| failed(e.to_string()))?;
    if verified {
        return Ok(());
    }
    // The verification ran and said no; the store context knows why.
    let reason = context
        .init(&store, leaf, &chain, |context| Ok(context.error()))
        .map(|error| error.to_string())
        .unwrap_or_else(|_| "unknown".to_owned());
    Err(failed(reason))
}

/// Verify a per-interface anchor is itself trusted by the global anchors.
///
/// Port of `check_trusted_ca`. Only the anchor file's **first** certificate is
/// checked, as Python checks `trusted_certs[0]`.
///
/// # Errors
///
/// The anchor does not chain to any of the roots, or a file is unreadable.
pub fn check_trusted_ca(
    root_ca_paths: &[impl AsRef<Path>],
    trusted_ca_path: &Path,
) -> Result<(), ConfigError> {
    let trusted = load_certs(trusted_ca_path)?;
    let roots = load_roots(root_ca_paths)?;
    let Some(leaf) = trusted.first() else {
        return Err(ConfigError(format!(
            "CONFIG: {} holds no certificate",
            repr(&trusted_ca_path.to_string_lossy()),
        )));
    };
    verify_chain(leaf, &roots, &[])
}

/// Verify the server certificate, its serial-bound SAN, and its private key.
///
/// Port of `check_certificate`. Three independent checks:
///
/// 1. the chain file's leaf, through its intermediates, chains to a root;
/// 2. the leaf's DNS SANs include `XYZ-<serial>`;
/// 3. the private key matches the leaf's public key.
///
/// # Errors
///
/// Any of the three, or a file that cannot be read or parsed.
pub fn check_certificate(
    root_ca_paths: &[impl AsRef<Path>],
    cert_path: &Path,
    key_path: &Path,
    serial_number: &str,
) -> Result<(), ConfigError> {
    let chain = load_certs(cert_path)?;
    let roots = load_roots(root_ca_paths)?;

    let Some((leaf, intermediates)) = chain.split_first() else {
        return Err(ConfigError(format!(
            "CONFIG: {} holds no certificate",
            repr(&cert_path.to_string_lossy()),
        )));
    };
    verify_chain(leaf, &roots, intermediates)?;

    // The uppercase `XYZ-` prefix is what the certificate generator actually
    // emits, verified against the shipped ExampleDeviceServer certificates --
    // not a guess at a convention.
    let expected = format!("XYZ-{serial_number}");
    let sans: Vec<String> = leaf
        .subject_alt_names()
        .map(|names| {
            names
                .iter()
                .filter_map(|name| name.dnsname().map(str::to_owned))
                .collect()
        })
        .unwrap_or_default();
    if !sans.contains(&expected) {
        return Err(ConfigError(format!(
            "certificate {} SAN does not include {}; found DNS SANs: {}",
            repr(&cert_path.to_string_lossy()),
            repr(&expected),
            repr_list(&sans),
        )));
    }

    // Compare the DER SubjectPublicKeyInfo of each, which works the same for
    // RSA, EC and the edwards curves without narrowing to a key type.
    let key_pem = std::fs::read(key_path).map_err(|error| {
        ConfigError(format!(
            "cannot read {}: {error}",
            repr(&key_path.to_string_lossy())
        ))
    })?;
    let key = openssl::pkey::PKey::private_key_from_pem(&key_pem).map_err(|error| {
        ConfigError(format!(
            "cannot load private key {}: {error}",
            repr(&key_path.to_string_lossy()),
        ))
    })?;
    let leaf_public = leaf
        .public_key()
        .and_then(|public| public.public_key_to_der())
        .map_err(|error| ConfigError(format!("cannot read the certificate's key: {error}")))?;
    let key_public = key
        .public_key_to_der()
        .map_err(|error| ConfigError(format!("cannot read the private key: {error}")))?;
    if leaf_public != key_public {
        return Err(ConfigError(format!(
            "private key {} does not match certificate {}",
            repr(&key_path.to_string_lossy()),
            repr(&cert_path.to_string_lossy()),
        )));
    }
    Ok(())
}

/// Fail fast on a mis-configured certificate set.
///
/// The order of the checks is the order `validate_startup_certs` performs them,
/// because the first failure is the one reported and an operator comparing the
/// two implementations should get the same diagnostic first.
///
/// # Errors
///
/// The configuration cannot be served; the message says why.
pub fn validate_startup_certs(args: &Args) -> Result<(), ConfigError> {
    if args.registry_disable_tls {
        // Nothing to validate: this is the deliberate plain-HTTP configuration.
        return Ok(());
    }

    if args.registry_certificate.is_empty() || args.registry_key.is_empty() {
        return Err(ConfigError(
            "CONFIG: TLS is enabled but --registryCertificate / --registryKey \
             were not supplied. Pass both, or run with --registryDisableTLS."
                .to_owned(),
        ));
    }

    // Both options are repeatable, so they must pair up. Checked before any
    // file is opened: a count mismatch is a usage error, and reporting it as a
    // missing file would name the wrong problem. Worded exactly as
    // `nmos_registry.py` words it -- `tests/config_parity.rs` runs both
    // binaries over the same argv and compares this line to the other's.
    if args.registry_certificate.len() != args.registry_key.len() {
        return Err(ConfigError(format!(
            "CONFIG: --registryCertificate was given {} time(s) but \
             --registryKey {} time(s); pass one key per certificate, in the \
             same order",
            args.registry_certificate.len(),
            args.registry_key.len(),
        )));
    }

    // Certificate then key within each pair, so a single-identity
    // configuration produces exactly the diagnostics it did before these
    // options became repeatable.
    for (certificate, key) in cli::identities(&args.registry_certificate, &args.registry_key) {
        for (role, path) in [
            ("--registryCertificate", certificate.as_path()),
            ("--registryKey", key.as_path()),
        ] {
            if !path.is_file() {
                return Err(ConfigError(format!(
                    "CONFIG: {role} is not accessible: {}",
                    repr(&path.to_string_lossy()),
                )));
            }
        }
    }

    let interface_cas: [(&str, &Vec<std::path::PathBuf>); 2] = [
        (
            "--registrationTrustedRootCA",
            &args.registration_trusted_root_ca,
        ),
        ("--queryTrustedRootCA", &args.query_trusted_root_ca),
    ];

    if interface_cas.iter().any(|(_, cas)| !cas.is_empty()) {
        if args.trusted_root_ca.is_empty() {
            return Err(ConfigError(
                "CONFIG: --trustedRootCA is required to validate the \
                 per-interface trusted root CAs"
                    .to_owned(),
            ));
        }
        for path in &args.trusted_root_ca {
            if !path.is_file() {
                return Err(ConfigError(format!(
                    "CONFIG: --trustedRootCA is not accessible: {}",
                    repr(&path.to_string_lossy()),
                )));
            }
        }
        for (role, cas) in interface_cas {
            for path in cas {
                if !path.is_file() {
                    return Err(ConfigError(format!(
                        "CONFIG: {role} is not accessible: {}",
                        repr(&path.to_string_lossy()),
                    )));
                }
                check_trusted_ca(&args.trusted_root_ca, path).map_err(|error| {
                    ConfigError(format!(
                        "CONFIG: {role} {} is not valid based on global \
                         --trustedRootCA: {error}",
                        repr(&path.to_string_lossy()),
                    ))
                })?;
            }
        }
    }

    if !args.trusted_root_ca.is_empty() {
        // Per identity: each must chain to a configured root and each key must
        // match its own certificate. Two identities of different types chain
        // to different roots in this PKI, so --trustedRootCA has to carry
        // both -- named here rather than at a peer's handshake.
        for (certificate, key) in cli::identities(&args.registry_certificate, &args.registry_key) {
            check_certificate(
                &args.trusted_root_ca,
                certificate.as_path(),
                key.as_path(),
                &args.registry_serial_number,
            )
            .map_err(|error| {
                ConfigError(format!(
                    "CONFIG: --registryCertificate is not valid: {error}"
                ))
            })?;
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use clap::Parser as _;
    use std::path::PathBuf;

    fn parse(extra: &[&str]) -> Args {
        let mut argv = vec!["nmos-registry"];
        argv.extend_from_slice(extra);
        Args::parse_from(argv)
    }

    fn certs() -> PathBuf {
        Path::new(env!("CARGO_MANIFEST_DIR"))
            .ancestors()
            .nth(3)
            .expect("crate is nested three deep under the repository root")
            .join("Certificates/build.0")
    }

    fn chain() -> String {
        certs()
            .join("pem/ExampleDeviceServer.ABC.SNX00000.chain.pem")
            .to_string_lossy()
            .into_owned()
    }

    fn key() -> String {
        certs()
            .join("key/ExampleDeviceServer.ABC.SNX00000.key")
            .to_string_lossy()
            .into_owned()
    }

    fn root() -> String {
        certs()
            .join("ExampleRootCA.pem")
            .to_string_lossy()
            .into_owned()
    }

    fn pki_available() -> bool {
        let present = Path::new(&chain()).is_file();
        if !present {
            eprintln!("skipping: the PKI under Certificates/build.0 is not present");
        }
        present
    }

    #[test]
    fn tls_without_a_certificate_refuses_to_start() {
        // The one that matters most: without this the registry would come up on
        // plain HTTP while its operator believed it was running TLS.
        let error = validate_startup_certs(&parse(&[])).expect_err("must refuse");
        assert!(
            error.to_string().contains("--registryCertificate"),
            "{error}"
        );
        assert!(
            error.to_string().contains("--registryDisableTLS"),
            "{error}"
        );
    }

    #[test]
    fn disabling_tls_skips_validation_entirely() {
        // The deliberate plain-HTTP configuration is not a misconfiguration.
        validate_startup_certs(&parse(&["--registryDisableTLS"])).expect("must not refuse");
    }

    #[test]
    fn a_certificate_path_that_does_not_exist_refuses_to_start() {
        if !pki_available() {
            return;
        }
        let error = validate_startup_certs(&parse(&[
            "--registryCertificate",
            "/nonexistent/cert.pem",
            "--registryKey",
            &key(),
        ]))
        .expect_err("must refuse");
        assert!(error.to_string().contains("not accessible"), "{error}");
    }

    #[test]
    fn a_missing_key_is_refused_even_when_the_certificate_is_present() {
        if !pki_available() {
            return;
        }
        let error = validate_startup_certs(&parse(&["--registryCertificate", &chain()]))
            .expect_err("must refuse");
        assert!(error.to_string().contains("--registryKey"), "{error}");
    }

    #[test]
    fn an_interface_anchor_requires_the_global_anchor() {
        if !pki_available() {
            return;
        }
        // "A mis-issued anchor would otherwise be discovered only when a
        // legitimate client was rejected."
        let error = validate_startup_certs(&parse(&[
            "--registryCertificate",
            &chain(),
            "--registryKey",
            &key(),
            "--registrationTrustedRootCA",
            &root(),
        ]))
        .expect_err("must refuse");
        assert!(error.to_string().contains("trustedRootCA"), "{error}");
    }

    #[test]
    fn a_complete_configuration_is_accepted() {
        if !pki_available() {
            return;
        }
        // Exercises all three of `check_certificate`'s checks against the real
        // PKI: the chain, the `XYZ-SNX00000` SAN, and the key match.
        validate_startup_certs(&parse(&[
            "--registryCertificate",
            &chain(),
            "--registryKey",
            &key(),
            "--registrySerialNumber",
            "SNX00000",
            "--trustedRootCA",
            &root(),
            "--registrationTrustedRootCA",
            &root(),
        ]))
        .expect("the shipped PKI should validate");
    }

    #[test]
    fn the_serial_must_appear_in_the_certificate_sans() {
        if !pki_available() {
            return;
        }
        // The serial binds the certificate to this device. A registry started
        // with someone else's serial would mint `aud` values it cannot honour.
        let error = validate_startup_certs(&parse(&[
            "--registryCertificate",
            &chain(),
            "--registryKey",
            &key(),
            "--registrySerialNumber",
            "SNX09999",
            "--trustedRootCA",
            &root(),
        ]))
        .expect_err("a mismatched serial must refuse");
        assert!(error.to_string().contains("XYZ-SNX09999"), "{error}");
        assert!(error.to_string().contains("SAN"), "{error}");
    }

    #[test]
    fn a_key_belonging_to_another_certificate_is_refused() {
        if !pki_available() {
            return;
        }
        // The EC identity's key against the RSA identity's certificate: both
        // are real and valid, and they do not go together.
        let other_key = certs().join("key/ExampleDeviceServer.ABC.SNX00000.ec.key");
        if !other_key.is_file() {
            eprintln!("skipping: no EC key in the PKI");
            return;
        }
        let error = validate_startup_certs(&parse(&[
            "--registryCertificate",
            &chain(),
            "--registryKey",
            &other_key.to_string_lossy(),
            "--registrySerialNumber",
            "SNX00000",
            "--trustedRootCA",
            &root(),
        ]))
        .expect_err("a mismatched key must refuse");
        assert!(
            error.to_string().contains("does not match certificate"),
            "{error}",
        );
    }

    #[test]
    fn an_anchor_that_does_not_chain_to_the_global_one_is_refused() {
        if !pki_available() {
            return;
        }
        // The product CA is a real intermediate, but it is not something the
        // *client* CA bundle vouches for on its own. Using the device leaf as
        // an "anchor" is the clearer case: it chains nowhere useful.
        let leaf_as_anchor = certs().join("pem/ExampleDeviceClient.ABC.SNX00000.chain.pem");
        if !leaf_as_anchor.is_file() {
            eprintln!("skipping: no client chain in the PKI");
            return;
        }
        let product = certs().join("ExampleProductCA.0.0.pem");
        if !product.is_file() {
            eprintln!("skipping: no product CA in the PKI");
            return;
        }
        // Global anchor is the *product* CA; the interface anchor is the root,
        // which the product CA does not vouch for.
        let error = validate_startup_certs(&parse(&[
            "--registryCertificate",
            &chain(),
            "--registryKey",
            &key(),
            "--registrySerialNumber",
            "SNX00000",
            "--trustedRootCA",
            &product.to_string_lossy(),
            "--registrationTrustedRootCA",
            &leaf_as_anchor.to_string_lossy(),
        ]))
        .expect_err("an unvouched anchor must refuse");
        assert!(
            error.to_string().contains("not valid based on global"),
            "{error}",
        );
    }
}
